src/Compilers/CSharp/Portable/Lowering/StateMachineRewriter/MethodToStateMachineRewriter.cs

```diff
@@ -434,6 +429,50 @@ private void AddVariableCleanup(ArrayBuilder<BoundExpression> cleanup, FieldSymb
             }
         }
 
+#nullable enable
+        protected BoundBlock GenerateAllHoistedLocalsCleanup()
```

Comment:
>GenerateAllHoistedLocalsCleanup [](http://example.com/codeflow?start=29&length=31)

This code looks very different from original implementation of ```GenerateHoistedLocalsCleanup```. Is it based on some other existing code? Also, isn't there a more direct and robust way to enumerate fields used for hoisting? I mean without going through locals, interpreting their replacements, etc. If not, perhaps we can build the list in advance, when the fields are added.